channels/offline-channel-js/src/OfflineBatchHandler.ts

```diff
@@ -0,0 +1,387 @@
+                                                    evt.attempCnt++;
+                                                    let sender = xhrOverride.sendPOST; // use transports
+                                                    let onCompleteCallback = (status: number, headers: {
+                                                        [headerName: string]: string;
+                                                    }, res?: string) => {
+                                                     
+                                                        if (status == 200) { // status code (e.g: invalid key) drop payload
+                                                            result.state = eBatchSendStatus.Complete;
+                                                            
+                                                        } else {
+                                                            let isRetriable = _isRetriable(evt, status);
+                                                            if (isRetriable) {
+                                                                result.state = eBatchSendStatus.Retry;
+                                                                _unloadProvider && _unloadProvider.addEvent(evt.id, evt, _itemCtx);
+
+                                                            } else {
+                                                                result.state = eBatchSendStatus.Drop;
+                                                            }
+                                                        }
+                                                        storeResolve(result);
+                                                        
+                                                    };
+                                                    return doAwaitResponse(sender(evt, onCompleteCallback, sync) as any, (res) => {
+                                                        // do nothing, because it should be resolved in callback
+                                                    });
+                                                }
+                                                // call resolve
+                                                result.state = eBatchSendStatus.Drop;
+                                                storeResolve(result);
+                                                
+                                            } catch (e) {
+                                                reject(e)
+                                            }
+                                        }
+                                        // this case should never happen because this function is called after we confirm that evt is available
+                                        // add resolve here in case
+                                        result.state = eBatchSendStatus.Complete;
+                                        storeResolve(result);
+
```

Comment:
Probably should return here, so that the `result.state` (if this code ever executes) doesn't get replaced with the `Failure` below.

While you "can" call the `resolve` (the `storeResolve`) of a promise more than once, and the code handles only really calling once -- we shouldn't rely on this behavior in case a future browser / Javascript framework starts treating that as an exception and throws.